### PR TITLE
Dev cnns new optimizers

### PR DESCRIPTION
**Pull request by Flowingsun007**

### CODE REVIEW

**Comment by Flowingsun007** [comment]
将原先基于参数：--model_update和方法：train_config.train.model_update_conf实现的optimizer、learning rate、warmup等替换成新的optimizer api。做了如下改动：
1.参数方面：去掉了--model_update 替换成--optimizer ; --mom改成--momentum其余几乎保持原样
2.原学习率设定(model_update_conf['learning_rate_decay'] ..)改成新flow.optimizer中的lr_scheduler，如：
flow.optimizer.CosineScheduler,PolynomialSchduler,PiecewiseScalingScheduler..
3.原warmup设定(model_update_conf['warmup_conf'])改成了flow.optimizer.warmup.linear(warmup_batches, 0)
4.原梯度截取设定(model_update_conf['clip_conf'] ..)改成了flow.optimizer.grad_clipping.by_global_norm(args.gradient_clipping) 
5.原各种优化器设定改成了flow.optimizer.SGD，Adam，AdamW，RMSProp

File: Classification/cnns/config.py [hunk 1]
```diff
@@ -95,7 +95,7 @@ def str2bool(v):
                         default='NHWC', help="NCHW or NHWC")
     parser.add_argument('--image-shape', type=int_list, default=[3, 224, 224],
                         help='the image shape feed into the network')
-    parser.add_argument('--label-smoothing', type=float, default=0.1, help='label smoothing factor')
+    parser.add_argument('--label_smoothing', type=float, default=0.0, help='label smoothing factor')
```
Comment (ShawnXuan) [inline Classification/cnns/config.py:98]: why change default value?

File: Classification/cnns/of_cnn_train_val.py [hunk 2]
```diff
@@ -80,10 +80,13 @@ def TrainNet():
     else:
         loss = flow.nn.sparse_softmax_cross_entropy_with_logits(labels, logits, name="softmax_loss")
 
-    loss = flow.math.reduce_mean(loss)
+    loss = flow.math.reduce_mean(loss)
```
Comment (ShawnXuan) [inline Classification/cnns/of_cnn_train_val.py:83]: blank space is not necessary.

File: Classification/cnns/optimizer_util.py [hunk 3]
```diff
@@ -13,104 +13,105 @@
...
+        ).minimize(loss)
 
 
 if __name__ == '__main__':
     import config as configs
     parser = configs.get_parser()
     args = parser.parse_args()
-    configs.print_args(args)
-    gen_model_update_conf(args)
+    configs.print_args(args)
```
Comment (ShawnXuan) [inline Classification/cnns/optimizer_util.py:117]: please try to fix stop sign here.

File: Classification/cnns/train.sh [hunk 4]
```diff
@@ -18,48 +19,52 @@ python3 of_cnn_train_val.py \
...
+#    --num_nodes=1 \
+#    --gpu_num_per_node=1 \
+#    --optimizer="sgd" \
+#    --momentum=0.875 \
+#    --learning_rate=0.001 \
+#    --loss_print_every_n_iter=1 \
+#    --batch_size_per_device=16 \
+#    --val_batch_size_per_device=10 \
+#    --num_epoch=10 \
+#    --model="resnet50"
```
Comment (ShawnXuan) [inline Classification/cnns/train.sh:41]: these changes are not necessary.

File: Classification/cnns/train.sh [hunk 5]
```diff
@@ -18,48 +19,52 @@ python3 of_cnn_train_val.py \
...
 
 # # training with imagenet
-# DATA_ROOT=/datasets/ImageNet/ofrecord
-# LOG_FOLDER=../logs
-# mkdir -p $LOG_FOLDER
-# LOGFILE=$LOG_FOLDER/resnet_training.log
+#  DATA_ROOT=/datasets/ImageNet/ofrecord
+#  LOG_FOLDER=../logs
+#  mkdir -p $LOG_FOLDER
+#  LOGFILE=$LOG_FOLDER/resnet_training.log
```
Comment (ShawnXuan) [inline Classification/cnns/train.sh:50]: these changes are not necessary.

File: Classification/cnns/train.sh [hunk 6]
```diff
@@ -18,48 +19,52 @@ python3 of_cnn_train_val.py \
...
+#      --gpu_num_per_node=8 \
+#      --optimizer="sgd" \
+#      --momentum=0.875 \
+#      --label_smoothing=0.1 \
+#      --learning_rate=1.024 \
+#      --loss_print_every_n_iter=100 \
+#      --batch_size_per_device=128 \
+#      --val_batch_size_per_device=100 \
+#      --num_epoch=90 \
+#      --model="resnet50" 2>&1 | tee ${LOGFILE}
```
Comment (ShawnXuan) [inline Classification/cnns/train.sh:67]: these changes are not necessary.

**Review by ShawnXuan — CHANGES_REQUESTED** [review]
Please review all changes, remove unnecessary ones, refer my comments.

**Review by ShawnXuan — COMMENTED** [review]
Did the regression tests cover all changes of optimizer API?

File: Classification/cnns/train.sh [hunk 7]
```diff
@@ -2,13 +2,14 @@
...
 
+
 # training with synthetic data
 python3 of_cnn_train_val.py \
     --num_examples=50 \
     --num_val_examples=50 \
     --num_nodes=1 \
     --gpu_num_per_node=1 \
-    --model_update="momentum" \
+    --optimizer="adam" \
```
Comment (ShawnXuan) [inline Classification/cnns/train.sh:12]: suggest to keep `momentum` as the default optimizer

File: Classification/cnns/train.sh (hunk 4 above)
Comment (ShawnXuan) [inline Classification/cnns/train.sh:41]: changes above are not necessary

File: Classification/cnns/train.sh (hunk 6 above)
Comment (ShawnXuan) [inline Classification/cnns/train.sh:67]: changes above are not necessary

File: Classification/cnns/train.sh (hunk 5 above)
Comment (ShawnXuan) [inline Classification/cnns/train.sh:50]: changes above are not necessary

**Review by ShawnXuan — CHANGES_REQUESTED** [review]
Keep a PR as clean as possible, there are a lot of useless changes which will disturb, mislead reviewer and waste their time.

File: Classification/cnns/train.sh (hunk 4 above)
Comment (ShawnXuan) [reply Classification/cnns/train.sh:41]: We can leave these comments unchanged in this PR.
or start a new PR to remove useless comment lines like here or below.

File: Classification/cnns/train.sh (hunk 7 above)
Comment (Flowingsun007) [reply Classification/cnns/train.sh:12]: momentum is no longer exists alone in new optimizer api, instead, 
a new way to use momentum is :  --optimizer="sdg"  --momentum=0.875

File: Classification/cnns/train.sh (hunk 4 above)
Comment (Flowingsun007) [reply Classification/cnns/train.sh:41]: > changes above are not necessary


I think its necessary and friendly to use ,considering we have three ways to training model:
1.training with synthetic data
2.training with mini-imagenet
2.training with imagenet

**Comment by Flowingsun007** [comment]
> 回归测试是否涵盖了优化程序API的所有更改？

**Comment by Flowingsun007** [comment]
> Did the regression tests cover all changes of optimizer API?

Yes, tests covered all changes api(including optimizers,lr scheduler,warmup)

File: Classification/cnns/train.sh (hunk 7 above)
Comment (ShawnXuan) [reply Classification/cnns/train.sh:12]: This `momentum` is the option of CNN's optimizer but not oneflow api, the default resnet50 optimizer is momentum(sgd + mom) but not `adam` .

File: Classification/cnns/train.sh (hunk 4 above)
Comment (ShawnXuan) [reply Classification/cnns/train.sh:41]: These are commented lines, I suggest to keep them unchanged, and start a new PR to handle them.
In the new PR, we can remove all commented lines, and update README about these three options.

**Review by ShawnXuan — CHANGES_REQUESTED** [review]

File: Classification/cnns/train.sh (hunk 7 above)
Comment (Flowingsun007) [reply Classification/cnns/train.sh]: ok,already modified

File: Classification/cnns/train.sh (hunk 4 above)
Comment (Flowingsun007) [reply Classification/cnns/train.sh:42]: ok,agree

**Review by ShawnXuan — APPROVED** [review]
